tests/auth/linking.node.test.ts

```diff
@@ -0,0 +1,186 @@
+import expect from "expect"
+
+import { LOCAL_IMPLEMENTATION } from "../../src/auth/local.js"
+import { createConsumer } from "../../src/auth/linking/consumer.js"
+import { createProducer } from "../../src/auth/linking/producer.js"
+import { EventEmitter } from "../../src/common/event-emitter.js"
+import { debug, setImplementations } from "../../src/setup.js"
+
+import type { Channel, ChannelOptions } from "../../src/auth/channel.js"
+
+type MessageData = string | ArrayBufferLike | Blob | ArrayBufferView
+
+/** Test implementation
+ * The goal of test suite it to test the interaction between producers and consumers. Delegation
+ * and the capablility to delegate are not tested.
+ * 
+ * The tests use an event emitter to emulate a networked communication channel. Delegating an account
+ * adds a username and a consumer DID to be linked to a list of producers. Linking an account adds a 
+ * username and consumer DID to a list of consumers.
+ * 
+ * If all consumers are linked at the end of a test, the consumers and producers should be the same. Errors
+ * or declined authorizations will set up different expectations.
+ * 
+ */
+
+describe("account linking", () => {
+  let channel: EventEmitter = new EventEmitter()
+  let producerAccounts: Record<string, string[]> = {}
+  let consumerAccounts: Record<string, string[]> = {}
+
+  before(() => {
+    const createChannel = async (options: ChannelOptions): Promise<Channel> => {
+      const { username, handleMessage } = options
+
+      const messageCallback = (data: MessageData) => { handleMessage(new MessageEvent(`${username}`, { data })) }
+      channel.addEventListener(`${username}`, messageCallback)
+
+      return {
+        send: (data) => channel.dispatchEvent(`${username}`, data),
+        close: () => channel.removeEventListener(`${username}`, messageCallback)
+      }
+    }
+
+    const checkCapability = async (username: string): Promise<boolean> => {
+      return true
+    }
+
+    const delegateAccount = async (username: string, audience: string): Promise<Record<string, unknown>> => {
+      producerAccounts[username] = producerAccounts[username] ?? []
+      producerAccounts[username] = [...producerAccounts[username], audience]
+
+      return { username, audience }
+    }
+
+    const linkDevice = async (data: Record<string, unknown>): Promise<void> => {
+      const { username, audience } = data as Record<string, string>
+
+      consumerAccounts[username] = consumerAccounts[username] ?? []
+      consumerAccounts[username] = [...consumerAccounts[username], audience]
+    }
+
+    setImplementations({
+      auth: {
+        ...LOCAL_IMPLEMENTATION.auth,
+        createChannel,
+        checkCapability,
+        delegateAccount,
+        linkDevice
+      }
+    })
```

Comment:
This is roughly what dependency injection will look like from a developer perspective. (With different function implementations, these ones are only good for the tests.)